### ISSUE TITLE
[v2 BUG]: MCC loss can give Nan

### ISSUE THREAD

**Issue by KnathanM** [chemprop/chemprop#846]
**Describe the bug**
`BinaryMCCLoss` can give a loss of `nan`.

**Example(s)**
The confusion matrix elements (TP, FP, TN, FN) are calculated as `targets * predictions`. At the start of training model predictions can happen to be negative, in which case some of the matrix elements can be negative. This leads `((TP + FP) * (TP + FN) * (TN + FP) * (TN + FN)).sqrt()` to possibly take the square root of a negative number and give `nan`. 

**Expected behavior**
I'm not sure what the expected behavior should be. Perhaps raw predictions should be forced to be positive (via a sigmoid?). Or the loss set to zero when the loss is `nan`?

**Additional context**
This was also a problem in v1 (#380). Our current implementation seems to be taken directly from [v1](https://github.com/chemprop/chemprop/blob/02c760216ec03488b69b41abe184589b9c996d74/chemprop/train/loss_functions.py#L87).

**Comment by ignaczgerg** [comment]
If anyone needs a quick solution for this, I just replaced the old `BinaryMCCLoss` class with this: 
```
@LossFunctionRegistry.register("binary-mcc")
class BinaryMCCLoss(LossFunction):
    def forward(
        self,
        preds: Tensor,
        targets: Tensor,
        mask: Tensor,
        weights: Tensor,
        lt_mask: Tensor,
        gt_mask: Tensor,
    ) -> Tensor:
        device = preds.device
        self.task_weights = self.task_weights.to(device)

        if not (0 <= preds.min() and preds.max() <= 1):  # assume logits
            preds = torch.sigmoid(preds) 

        L = self._calc_unreduced_loss(preds, targets.long(), mask, weights, lt_mask, gt_mask)
        L = L * self.task_weights

        return L.mean()

    def _calc_unreduced_loss(self, preds, targets, mask, weights, *args) -> Tensor:
        device = preds.device
        if weights is None:
            weights = torch.ones_like(targets).to(device)

        TP = (targets * preds * weights * mask).sum(0, keepdim=True)
        FP = ((1 - targets) * preds * weights * mask).sum(0, keepdim=True)
        TN = ((1 - targets) * (1 - preds) * weights * mask).sum(0, keepdim=True)
        FN = (targets * (1 - preds) * weights * mask).sum(0, keepdim=True)

        numerator = TP * TN - FP * FN
        denominator = ((TP + FP) * (TP + FN) * (TN + FP) * (TN + FN)).sqrt()

        MCC = numerator / (denominator + 1e-8)

        return 1 - MCC
```

1. I added the old `MccMixin` loss calculation directly to the `BinaryMCCLoss`'s forward function. 
2. I also added an edge case, `if weights is None`. I think this part can potentially be dangerous later on, so it might better to add a logger warning. 
3. Based on the comments above, I also added 10^-8 to the denominator for numerical stability. 

Even though the old tests are running fine, I did not test it thoroughly. I also run a few learning tasks with it (one toy example and some real data) and the results looked good. 

I hope this might help anyone!
